dcicutils/license_utils.py

```diff
@@ -762,648 +780,230 @@ def __init__(self, message=None, unacceptable_licenses=None):
+    else:
+        filename = os.path.join(policy_dir, policy_name + ".jsonc")
+        if not os.path.exists(filename):
+            if error_if_missing:
+                raise ValueError(f"No such policy: {policy_name!r}")
+            else:
+                return None
+        data = read_license_policy_file(filename)
+        POLICY_DATA_CACHE[policy_name] = data
+        return data
 
 
-@LicenseCheckerRegistry.register_checker('c4-python-infrastructure')
-class C4PythonInfrastructureLicenseChecker(C4InfrastructureLicenseChecker):
+def find_or_create_license_class(*, policy_name: str, policy_dir: str, for_env,
+                                 # This next argument should never be passed explicitly by callers other than
+                                 # recursive calls to this function. -kmp 28-Sep-2023
+                                 _creation_attmpts_in_progress=None):
+    """
+    Define a policy class given a policy name (like 'c4-infrastructure').
+    """
+    _creation_attmpts_in_progress = _creation_attmpts_in_progress or []
+    existing_checker = LicenseCheckerRegistry.find_checker(checker_name=policy_name)
+    if existing_checker:
+        return existing_checker
+    elif policy_name in _creation_attmpts_in_progress:
+        raise ValueError(f"Circular reference to {policy_name} detected"
+                         f" while creating {conjoined_list(_creation_attmpts_in_progress)}.")
+    _creation_attmpts_in_progress.append(policy_name)
+    license_checker_class_name = to_camel_case(policy_name) + "LicenseChecker"
+    policy_data = find_policy_data(policy_name, policy_dir=policy_dir)
+    inherits_from = policy_data.get('inherits_from')
+    if not isinstance(inherits_from, list):
+        raise ValueError(f'Policy {policy_name!r} needs "inherits_from": [...parent names...],'
+                         f' which may be empty but must be specified.')
+    license_frameworks = policy_data.get('LICENSE_FRAMEWORKS')
+    if license_frameworks == "ALL":
+        policy_data['LICENSE_FRAMEWORKS'] = LicenseFrameworkRegistry.all_framework_names()
+    parent_classes = [find_or_create_license_class(policy_name=parent_name, policy_dir=policy_dir, for_env=for_env,
+                                                   _creation_attmpts_in_progress=_creation_attmpts_in_progress)
+                      for parent_name in inherits_from]
+    defaulted_policy_data = default_policy_data(policy_name=policy_name, policy_data=policy_data,
+                                                parent_classes=parent_classes)
+    new_class = type(license_checker_class_name,
+                     (*parent_classes, LicenseChecker),
+                     {'_policy_data': policy_data, **defaulted_policy_data})
+    new_class.__doc__ = policy_data.get("description") or f'License policy {policy_name} needs a "description".'
+    assert isinstance(new_class, type) and issubclass(new_class, LicenseChecker)  # Sigh. PyCharm can't figure this out
```

Comment:
Yeah, but the reasons are obscure, so I expanded the comment:
```
# Sigh. PyCharm can't figure this out type fact out, even with a type hint on the above assignment to new_class,
# such as 'new_class: Type[LicenseChecker] = ...'. That should have worked. Putting in an assert was the only way
# I could find to convince PyCharm of the truth. I don't expect this assertion to ever fail. It's just an artifact
# to prevent ugly browser highlighting. I'll try to arrange a bug report for them. -kmp 29-Sep-2023
```